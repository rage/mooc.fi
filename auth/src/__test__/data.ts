import { NextFunction, Request, Response } from "express"

export const testProfile = {
  edu_person_affiliation: "",
  edu_person_principal_name: "test",
  schac_home_organization: "helsinki.fi",
  organizational_unit: "organizational_unit",
  schac_personal_unique_code:
    "urn:schac:personalUniqueCode:int:studentID:helsinki.fi:121345678",
  display_name: "test",
  given_name: "test",
  mail: "test@mail.com",
}

export const req = {
  body: {},
  query: {
    RelayState: JSON.stringify({
      language: "fi",
      provider: "hy",
      action: "sign-in",
    }),
  },
  params: {},
  login: jest.fn().mockImplementation((_: any, fn: Function) => fn(null)),
  logout: jest.fn(),
} as unknown as Request
export const res = {
  redirect: jest.fn().mockReturnThis(),
  locals: {},
  setMOOCCookies: jest.fn().mockReturnThis(),
} as unknown as Response
export const next = jest.fn() as NextFunction

export const okTokenResponse = {
  data: {
    status: 200,
    success: true,
    tmc_token: "tmc_token",
    access_token: "access_token",
    admin: false,
  },
}
