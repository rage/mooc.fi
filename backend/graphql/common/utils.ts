import { ApolloError, UserInputError } from "apollo-server-express"

import { Prisma } from "@prisma/client"
import {
  Organization,
  OrganizationRole,
  User,
  UserOrganization,
} from "@prisma/client"

import { Context } from "../../context"
import {
  emptyOrNullToUndefined,
  err,
  isDefined,
  isNullOrUndefined,
  ok,
  Result,
} from "../../util"
import { ConfigurationError, ConflictError } from "./errors"

export const buildUserSearch = (
  search?: string | null,
): Prisma.UserWhereInput => {
  if (isNullOrUndefined(search)) {
    return {}
  }
  return {
    OR: [
      {
        first_name: { contains: search, mode: "insensitive" },
      },
      {
        last_name: { contains: search, mode: "insensitive" },
      },
      {
        username: { contains: search, mode: "insensitive" },
      },
      {
        email: { contains: search, mode: "insensitive" },
      },
      {
        user_organizations: {
          some: {
            organizational_email: { contains: search, mode: "insensitive" },
          },
        },
      },
      {
        student_number: { contains: search },
      },
      {
        real_student_number: { contains: search },
      },
      {
        upstream_id: Number(search) ?? undefined,
      },
    ],
  }
}

interface ConvertPaginationInput {
  first?: number | null
  last?: number | null
  before?: string | null
  after?: string | null
  skip?: number | null
}

interface ConvertPaginationOptions {
  field?: string
}

interface ConvertPaginationOutput {
  skip?: number
  cursor?: { [key: string]: string }
  take?: number
}

export const convertPagination = (
  { first, last, before, after, skip }: ConvertPaginationInput,
  options?: ConvertPaginationOptions,
): ConvertPaginationOutput => {
  const skipValue = skip || 0
  const { field = "id" } = options || {}

  if (!first && !last) {
    throw new Error("first or last must be defined")
  }

  return {
    skip: isDefined(before) ? skipValue + 1 : skipValue,
    take: isDefined(last) ? -(last ?? 0) : isDefined(first) ? first : 0,
    cursor: isDefined(before)
      ? { [field]: before }
      : isDefined(after)
      ? { [field]: after }
      : undefined,
  }
}

export const getCourseOrAliasBySlug =
  (ctx: Context) => async (slug: string) => {
    let course = await ctx.prisma.course.findUnique({
      where: { slug },
    })

    if (!course) {
      course = await ctx.prisma.courseAlias
        .findUnique({
          where: {
            course_code: slug,
          },
        })
        .course()

      if (!course) {
        throw new UserInputError("course or course alias not found", {
          argumentName: "slug",
        })
      }
    }

    return course
  }

export const getCourseOrCompletionHandlerCourse =
  (ctx: Context) =>
  async ({ id, slug }: Prisma.CourseWhereUniqueInput) => {
    if (!id && !slug) {
      throw new UserInputError("must provide id and/or slug", {
        argumentName: ["id", "slug"],
      })
    }

    const course = await ctx.prisma.course.findUnique({
      where: {
        id,
        slug,
      },
      include: {
        completions_handled_by: true,
      },
    })

    return course?.completions_handled_by ?? course
  }

export const checkEmailValidity = (
  email?: string | null,
  requiredPattern?: string | null,
): Result<boolean, Error> => {
  if (!email) {
    return err(
      new UserInputError(
        "no email specified and no email found in user profile",
      ),
    )
  }

  if (requiredPattern) {
    if (!email.match(requiredPattern)) {
      return err(
        new UserInputError(
          "given email does not fulfill organization email requirements",
        ),
      )
    }
  }

  return ok(true)
}

interface JoinUserOrganizationOptions {
  ctx: Context
  user: User
  organization: Organization
  organizational_email?: string | null
  organizational_identifier?: string | null
  redirect?: string | null
  language?: string | null
}

export const joinUserOrganization = async ({
  ctx,
  user,
  organization,
  organizational_email,
  organizational_identifier,
}: JoinUserOrganizationOptions) => {
  const exists = await ctx.prisma.userOrganization.findFirst({
    where: {
      user_id: user.id,
      organization_id: organization.id,
    },
    select: {
      id: true,
    },
  })

  if (exists) {
    return err(
      new ConflictError("this user/organization relation already exists"),
    )
  }

  const { required_confirmation } = organization

  const currentDate = new Date()

  const userOrganization = await ctx.prisma.userOrganization.create({
    data: {
      user: { connect: { id: user.id } },
      organization: {
        connect: { id: organization.id },
      },
      role: OrganizationRole.Student,
      organizational_email: emptyOrNullToUndefined(organizational_email),
      organizational_identifier: emptyOrNullToUndefined(
        organizational_identifier,
      ),
      ...(!required_confirmation && {
        confirmed: true,
        confirmed_at: currentDate,
      }),
      // we assume that the consent is given in the frontend
      consented: true,
    },
  })

  return ok(userOrganization)
}

interface CancelEmailDeliveriesOptions {
  ctx: Context
  userOrganization: UserOrganization
  organization: Organization
  email?: string
  errorMessage?: string
}
export const cancelEmailDeliveries = async ({
  ctx,
  userOrganization,
  organization,
  email,
  errorMessage,
}: CancelEmailDeliveriesOptions) => {
  const { count } = await ctx.prisma.emailDelivery.updateMany({
    where: {
      user_id: userOrganization.user_id,
      email,
      email_template_id: organization.join_organization_email_template_id,
      organization_id: organization.id,
      sent: false,
      error: false,
    },
    data: {
      error: { set: true },
      error_message: errorMessage ?? `Email delivery canceled at ${new Date()}`,
    },
  })

  if (count) {
    ctx.logger.info(
      `Expired ${count} user organization join email deliver${
        count > 1 ? "ies" : "y"
      } still in the send queue; updated`,
    )
  }
}

interface CreateUserOrganizationJoinConfirmationOptions {
  ctx: Context
  user: User
  organization: Organization
  userOrganization: UserOrganization
  email: string
  redirect?: string | null
  language?: string | null
}

export const createUserOrganizationJoinConfirmation = async ({
  ctx,
  user,
  organization,
  userOrganization,
  email,
  redirect,
  language,
}: CreateUserOrganizationJoinConfirmationOptions) => {
  const { join_organization_email_template_id } = organization

  /*const isValid = checkEmailValidity(email, required_organization_email)

  if (isValid.isErr()) {
    return isValid
  }
  */

  if (!join_organization_email_template_id) {
    return err(
      new ConfigurationError(
        "no email template associated with this organization",
      ),
    )
  }

  const currentDate = new Date()

  try {
    const userOrganizationJoinConfirmation =
      await ctx.prisma.userOrganizationJoinConfirmation.create({
        data: {
          user_organization: { connect: { id: userOrganization.id } },
          email,
          email_delivery: {
            create: {
              user_id: user.id,
              email,
              email_template_id: join_organization_email_template_id,
              organization_id: organization.id,
              sent: false,
              error: false,
            },
          },
          redirect,
          language,
          expires_at: new Date(currentDate.getTime() + 4 * 60 * 60 * 1000), // 4 hours for now
        },
      })

    return ok(userOrganizationJoinConfirmation)
  } catch (e: unknown) {
    if (e instanceof Error) return err(e)

    return err(
      new ApolloError((e as string) ?? "Unknown error creating confirmation"),
    )
  }
}
