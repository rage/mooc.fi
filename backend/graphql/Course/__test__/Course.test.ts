import { createReadStream } from "fs"
import { gql } from "graphql-request"
import { getTestContext, fakeTMCCurrent } from "../../../tests/__helpers"
import {
  //normalUser,
  normalUserDetails,
  //adminUser,
  adminUserDetails,
} from "../../../tests/data"
import { seed } from "../../../tests/data/seed"

import { Course } from "@prisma/client"
import { orderBy } from "lodash"
import { mocked } from "ts-jest/utils"
import { omit } from "lodash"

jest.mock("../../../services/kafkaProducer")
import KafkaProducer from "../../../services/kafkaProducer"

const ctx = getTestContext()
const tmc = fakeTMCCurrent({
  "Bearer normal": [200, normalUserDetails],
  "Bearer admin": [200, adminUserDetails],
})

const courseQuery = gql`
  query course($id: ID, $slug: String, $language: String) {
    course(id: $id, slug: $slug, language: $language) {
      id
      name
      slug
      description
      link
    }
  }
`

const fullCourseQuery = gql`
  query course(
    $id: ID
    $slug: String
    $language: String
    $includeDeletedExercises: Boolean
  ) {
    course(id: $id, slug: $slug, language: $language) {
      id
      name
      slug
      ects
      order
      study_module_order
      teacher_in_charge_name
      teacher_in_charge_email
      support_email
      start_date
      end_date
      tier
      photo {
        id
        compressed
        compressed_mimetype
        uncompressed
        uncompressed_mimetype
      }
      promote
      start_point
      hidden
      study_module_start_point
      status
      exercises(includeDeleted: $includeDeletedExercises) {
        id
        name
        deleted
      }
      course_translations {
        id
        name
        language
        description
        link
      }
      open_university_registration_links {
        id
        course_code
        language
        link
      }
      study_modules {
        id
      }
      course_variants {
        id
        slug
        description
      }
      course_aliases {
        id
        course_code
      }
      inherit_settings_from {
        id
      }
      completions_handled_by {
        id
      }
      has_certificate
      user_course_settings_visibilities {
        id
        language
      }
      upcoming_active_link
      automatic_completions
      automatic_completions_eligible_for_ects
      exercise_completions_needed
      points_needed
    }
  }
`

const coursesQuery = gql`
  query AllCourses(
    $language: String
    $orderBy: CourseOrderByInput
    $search: String
    $hidden: Boolean
    $handledBy: String
  ) {
    courses(
      orderBy: $orderBy
      language: $language
      search: $search
      hidden: $hidden
      handledBy: $handledBy
    ) {
      id
      slug
      name
      order
      study_module_order
      photo {
        id
        compressed
        uncompressed
      }
      promote
      status
      start_point
      study_module_start_point
      hidden
      description
      link
      upcoming_active_link
      study_modules {
        id
        slug
      }
      course_translations {
        id
        language
        name
      }
      user_course_settings_visibilities {
        id
        language
      }
    }
  }
`

const courseExistsQuery = gql`
  query courseExists($slug: String!) {
    course_exists(slug: $slug)
  }
`

const createCourseMutation = gql`
  mutation createCourse($course: CourseCreateArg!) {
    addCourse(course: $course) {
      id
      name
      slug
      ects
      order
      study_module_order
      teacher_in_charge_name
      teacher_in_charge_email
      support_email
      start_date
      end_date
      tier
      photo {
        id
        original
        compressed
        compressed_mimetype
        uncompressed
        uncompressed_mimetype
      }
      promote
      start_point
      hidden
      study_module_start_point
      status
      course_translations {
        id
        name
        language
        description
        link
      }
      open_university_registration_links {
        id
        course_code
        language
        link
      }
      study_modules {
        id
      }
      course_variants {
        id
        slug
        description
      }
      course_aliases {
        id
        course_code
      }
      inherit_settings_from {
        id
      }
      completions_handled_by {
        id
      }
      has_certificate
      user_course_settings_visibilities {
        id
        language
      }
      upcoming_active_link
      automatic_completions
      automatic_completions_eligible_for_ects
      exercise_completions_needed
      points_needed
    }
  }
`

const updateCourseMutation = gql`
  mutation updateCourse($course: CourseUpsertArg!) {
    updateCourse(course: $course) {
      id
      name
      slug
      ects
      order
      study_module_order
      teacher_in_charge_name
      teacher_in_charge_email
      support_email
      start_date
      end_date
      tier
      photo {
        id
        original
        compressed
        compressed_mimetype
        uncompressed
        uncompressed_mimetype
      }
      promote
      start_point
      hidden
      study_module_start_point
      status
      course_translations {
        id
        name
        language
        description
        link
      }
      open_university_registration_links {
        id
        course_code
        language
        link
      }
      study_modules {
        id
      }
      course_variants {
        id
        slug
        description
      }
      course_aliases {
        id
        course_code
      }
      inherit_settings_from {
        id
      }
      completions_handled_by {
        id
      }
      has_certificate
      user_course_settings_visibilities {
        id
        language
      }
      upcoming_active_link
      automatic_completions
      automatic_completions_eligible_for_ects
      exercise_completions_needed
      points_needed
    }
  }
`

const deleteCourseMutation = gql`
  mutation deleteCourse($id: ID, $slug: String) {
    deleteCourse(id: $id, slug: $slug) {
      id
    }
  }
`

const handlerCoursesQuery = gql`
  query handlerCourses {
    handlerCourses {
      id
      handles_completions_for {
        id
      }
    }
  }
`

const courseCompletionsQuery = gql`
  query courseCompletions(
    $slug: String
    $user_id: String
    $user_upstream_id: Int
  ) {
    course(slug: $slug) {
      id
      completions(user_id: $user_id, user_upstream_id: $user_upstream_id) {
        id
        user {
          id
          username
        }
        completion_language
        email
        user_upstream_id
      }
    }
  }
`
// study_modules may be returned in any order, let's just sort them so snapshots are equal
const sortStudyModules = (course: any) => {
  if (!course?.study_modules) {
    return course
  }

  return {
    ...course,
    study_modules: orderBy(course.study_modules, ["id"], ["asc"]),
  }
}

const sortExercises = (course: any) => {
  if (!course?.exercises) {
    return course
  }

  return {
    ...course,
    exercises: orderBy(course.exercises, ["id"]),
  }
}

describe("Course", () => {
  afterAll(() => jest.clearAllMocks())

  describe("model", () => {
    beforeAll(() => tmc.setup())
    afterAll(() => tmc.teardown())

    describe("completions", () => {
      beforeEach(async () => {
        await seed(ctx.prisma)
      })

      it("errors on no admin", async () => {
        return ctx.client
          .request(
            courseCompletionsQuery,
            {
              slug: "course1",
              user_upstream_id: 1,
            },
            {
              Authorization: "Bearer normal",
            },
          )
          .then(() => fail())
          .catch(({ response }) => {
            expect(response.errors?.length).toBe(1)
            expect(response.errors[0].message).toContain("Not authorized")
            expect(response.status).toBe(200)
          })
      })

      it("errors with no user_id or user_upstream_id", async () => {
        return ctx.client
          .request(
            courseCompletionsQuery,
            {
              slug: "course1",
            },
            {
              Authorization: "Bearer admin",
            },
          )
          .then(() => fail())
          .catch(({ response }) => {
            expect(response.errors?.length).toBe(1)
            expect(response.errors[0].message).toContain(
              "needs user_id or user_upstream_id",
            )
            expect(response.status).toBe(200)
          })
      })

      it("works with user_id", async () => {
        const res = await ctx.client.request(
          courseCompletionsQuery,
          {
            slug: "course1",
            user_id: "20000000000000000000000000000103",
          },
          {
            Authorization: "Bearer admin",
          },
        )

        expect({
          ...res,
          completions: orderBy(res.completions, "id"),
        }).toMatchSnapshot()
      })

      it("works with user_upstream_id", async () => {
        const res = await ctx.client.request(
          courseCompletionsQuery,
          {
            slug: "course1",
            user_upstream_id: 1,
          },
          {
            Authorization: "Bearer admin",
          },
        )

        expect({
          ...res,
          completions: orderBy(res.completions, "id"),
        }).toMatchSnapshot()
      })

      it("shouldn't return anything with non-existent user", async () => {
        const res = await ctx.client.request(
          courseCompletionsQuery,
          {
            slug: "course1",
            user_upstream_id: 4939298,
          },
          {
            Authorization: "Bearer admin",
          },
        )

        expect({
          ...res,
          completions: orderBy(res.completions, "id"),
        }).toMatchSnapshot()
      })
    })
  })

  describe("queries", () => {
    beforeAll(() => tmc.setup())
    afterAll(() => tmc.teardown())

    describe("course", () => {
      let createdCourses: Course[] | null = null

      beforeEach(async () => {
        createdCourses = (await seed(ctx.prisma)).courses
      })

      afterEach(() => (createdCourses = null))

      describe("normal user", () => {
        beforeEach(() =>
          ctx!.client.setHeader("Authorization", "Bearer normal"),
        )

        it("should error on no parameters", async () => {
          try {
            await ctx.client.request(courseQuery)
            fail()
          } catch {}
        })

        it("returns course on id and slug", async () => {
          const resId = await ctx.client.request(courseQuery, {
            id: createdCourses![0].id,
          })
          const resSlug = await ctx.client.request(courseQuery, {
            slug: "course1",
          })

          ;[resId, resSlug].map((res) =>
            // had sortStudyModules
            expect(res.course).toMatchSnapshot({
              id: expect.any(String),
            }),
          )
        })

        it("returns correct language", async () => {
          const res = await ctx.client.request(courseQuery, {
            slug: "course1",
            language: "en_US",
          })

          // had sortStudyModules
          expect(res.course).toMatchSnapshot({
            id: expect.any(String),
          })
        })

        it("should return null on non-existent language", async () => {
          const res = await ctx.client.request(courseQuery, {
            slug: "course1",
            language: "sv_SE",
          })

          expect(res).toEqual({ course: null })
        })

        it("should error on invalid id and slug", async () => {
          try {
            await ctx.client.request(courseQuery, {
              id: new Array(33).join("1"),
            })
            fail()
          } catch {}
          try {
            await ctx.client.request(courseQuery, {
              slug: "invalid",
            })
            fail()
          } catch {}
        })
      })

      describe("admin", () => {
        beforeEach(() => ctx!.client.setHeader("Authorization", "Bearer admin"))

        it("returns full course on id and slug", async () => {
          const resId = await ctx.client.request(fullCourseQuery, {
            id: createdCourses![0].id,
          })
          const resSlug = await ctx.client.request(fullCourseQuery, {
            slug: "course1",
          })

          ;[resId, resSlug].map((res) =>
            expect(sortExercises(sortStudyModules(res.course))).toMatchSnapshot(
              {
                id: expect.any(String),
              },
            ),
          )
        })

        it("should include deleted exercises if specified", async () => {
          const res = await ctx.client.request(fullCourseQuery, {
            slug: "course1",
            includeDeletedExercises: true,
          })

          expect(sortExercises(sortStudyModules(res.course))).toMatchSnapshot()
        })
      })
    })

    describe("courses", () => {
      beforeEach(async () => {
        await seed(ctx.prisma)
      })

      it("returns courses", async () => {
        const res = await ctx.client.request(coursesQuery)

        expect(
          orderBy(res.courses.map(sortStudyModules), ["id"]),
        ).toMatchSnapshot()
      })

      it("returns courses ordered", async () => {
        for (const order of ["asc", "desc"]) {
          const res = await ctx.client.request(coursesQuery, {
            orderBy: { name: order },
          })

          expect(res.courses.map(sortStudyModules)).toMatchSnapshot(
            `courses-order-${order}`,
          )
        }
      })

      it("returns courses filtered by language", async () => {
        for (const language of ["fi_FI", "en_US", "bogus"]) {
          const res = await ctx.client.request(coursesQuery, {
            language,
          })

          expect(
            orderBy(res.courses?.map(sortStudyModules), ["id"]),
          ).toMatchSnapshot(`courses-language-${language}`)
        }
      })

      const searchTest: Array<[string, string[]]> = [
        ["course1", ["course1"]],
        ["teacher", ["course1", "course2"]],
        ["teacher1", ["course1"]],
        ["teacher2", ["course2"]],
        ["teacher3", []],
        ["e@mail", ["course1", "course2"]],
        ["en_US", ["course1"]],
        ["se_SE", []],
      ]

      it("returns search results", async () => {
        for (const [search, expected] of searchTest) {
          const res = await ctx.client.request(coursesQuery, {
            search,
          })

          const resultSlugs = res.courses?.map((c: Course) => c.slug).sort()

          expect(resultSlugs).toEqual(expected.sort())
        }
      })

      it("filters hidden", async () => {
        for (const hidden of [true, false, null]) {
          const res = await ctx.client.request(coursesQuery, {
            hidden,
          })

          expect(res.courses?.map((c: Course) => c.id).sort()).toMatchSnapshot(
            `courses-hidden-${hidden || "null"}`,
          )
        }
      })

      it("filters handledBy", async () => {
        for (const handledBy of ["handler", "foo"]) {
          const res = await ctx.client.request(coursesQuery, {
            handledBy,
          })

          expect(res.courses?.map((c: Course) => c.id).sort()).toMatchSnapshot(
            `courses-hidden-${handledBy}`,
          )
        }
      })
    })

    describe("course_exists", () => {
      beforeEach(async () => {
        await seed(ctx.prisma)
        ctx!.client.setHeader("Authorization", "Bearer normal")
      })

      it("returns true on existing course", async () => {
        const res = await ctx.client.request(courseExistsQuery, {
          slug: "course1",
        })

        expect(res).toEqual({ course_exists: true })
      })

      it("returns false on non-existing course", async () => {
        const res = await ctx.client.request(courseExistsQuery, {
          slug: "bogus",
        })

        expect(res).toEqual({ course_exists: false })
      })
    })

    describe("handlerCourses", () => {
      beforeEach(async () => {
        await seed(ctx.prisma)
        ctx!.client.setHeader("Authorization", "Bearer admin")
      })

      it("returns correctly", async () => {
        const res = await ctx.client.request(handlerCoursesQuery)

        expect(res.handlerCourses).toMatchSnapshot()
      })

      it("errors with non-admin", async () => {
        ctx!.client.setHeader("Authorization", "Bearer normal")
        try {
          await ctx!.client.request(handlerCoursesQuery)
          fail()
        } catch {}
      })
    })
  })

  describe("mutations", () => {
    beforeAll(() => tmc.setup())
    afterAll(() => tmc.teardown())

    const getNewCourse = () => ({
      name: "new1",
      slug: "new1",
      start_date: "01/01/1900",
      teacher_in_charge_email: "e@mail.com",
      teacher_in_charge_name: "teacher",
      study_modules: [{ id: "00000000000000000000000000000101" }],
      course_translations: [
        {
          description: "description_en_US",
          language: "en_US",
          name: "name_en_US",
        },
      ],
      course_variants: [
        {
          slug: "variant1",
          description: "variant1",
        },
      ],
      course_aliases: [
        {
          course_code: "alias1",
        },
      ],
      inherit_settings_from: "00000000000000000000000000000002",
      completions_handled_by: "00000000000000000000000000000002",
      user_course_settings_visibilities: [{ language: "en_US" }],
      new_photo: createReadStream(__dirname + "/../../../tests/data/image.gif"),
    })

    const getUpdateCourse = () => ({
      // id: "00000000000000000000000000000002",
      name: "updated course1",
      slug: "course1",
      new_slug: "updated_course1",
      start_date: "02/01/1900",
      end_date: "12/30/2100",
      teacher_in_charge_email: "updated-e@mail.com",
      teacher_in_charge_name: "updated teacher",
      course_translations: [
        {
          id: "00000000-0000-0000-0000-000000000011",
          description: "course1_updated_description_en_US",
          language: "en_US",
          name: "course1_en_US",
          link: "http://link.com",
        },
        {
          description: "course1_added_description_se_SE",
          language: "se_SE",
          name: "course1_se_SE",
          link: "http:/link.se.com",
        },
      ],
      study_modules: [
        {
          id: "00000000-0000-0000-0000-000000000101",
        },
      ],
      course_variants: [
        {
          slug: "variant1",
          description: "variant1",
        },
      ],
      course_aliases: [
        {
          course_code: "alias1",
        },
      ],
      inherit_settings_from: "00000000000000000000000000000001",
      completions_handled_by: "00000000000000000000000000000001",
      user_course_settings_visibilities: [{ language: "en_US" }],
      photo: "00000000000000000000000000001101",
      new_photo: createReadStream(__dirname + "/../../../tests/data/image.gif"),
    })

    const expectedAddedCourse = {
      id: expect.any(String),
      course_translations: [
        {
          id: expect.any(String),
        },
      ],
      course_variants: [
        {
          id: expect.any(String),
        },
      ],
      course_aliases: [
        {
          id: expect.any(String),
        },
      ],
      photo: {
        original: expect.stringContaining("image.gif"),
        compressed: expect.stringContaining("webp"),
        uncompressed: expect.stringContaining("jpeg"),
        id: expect.any(String),
      },
      user_course_settings_visibilities: [
        {
          id: expect.any(String),
        },
      ],
    }

    const expectedUpdatedCourse = {
      id: "00000000-0000-0000-0000-000000000002",
      study_modules: [
        {
          id: expect.stringMatching("00000000-0000-0000-0000-000000000101"),
        },
      ],
      course_translations: [
        {
          id: expect.any(String),
        },
        {
          id: expect.any(String),
        },
      ],
      course_variants: [
        {
          id: expect.any(String),
        },
      ],
      course_aliases: [
        {
          id: expect.any(String),
        },
      ],
      photo: {
        original: expect.stringContaining("image.gif"),
        compressed: expect.stringContaining("webp"),
        uncompressed: expect.stringContaining("jpeg"),
        id: expect.any(String),
      },
      user_course_settings_visibilities: [
        {
          id: expect.any(String),
        },
      ],
    }

    type TestCase = [
      string,
      {
        data: object
        expected: object
      },
    ]

    describe("addCourse", () => {
      beforeEach(async () => {
        await seed(ctx.prisma)
        ctx!.client.setHeader("Authorization", "Bearer admin")
        mocked(KafkaProducer).mockClear()
      })

      const cases: TestCase[] = [
        [
          "full",
          {
            data: getNewCourse(),
            expected: expectedAddedCourse,
          },
        ],
        [
          "with no course_translations",
          {
            data: omit(getNewCourse(), "course_translations"),
            expected: omit(expectedAddedCourse, "course_translations"),
          },
        ],
        [
          "with no study_modules",
          {
            data: omit(getNewCourse(), "study_modules"),
            expected: expectedAddedCourse,
          },
        ],
        [
          "with no photo",
          {
            data: omit(getNewCourse(), "new_photo"),
            expected: omit(expectedAddedCourse, "photo"),
          },
        ],
      ]

      test.each(cases)("creates a course %s", async (_, { data, expected }) => {
        const res = await ctx!.client.request(createCourseMutation, {
          course: data,
        })

        expect(res).toMatchSnapshot({
          addCourse: expected,
        })

        expect(KafkaProducer).toHaveBeenCalledTimes(1)

        const createdCourse = await ctx.prisma.course.findFirst({
          where: { slug: "new1" },
        })
        expect(
          mocked(KafkaProducer).mock.instances[0].queueProducerMessage,
        ).toHaveBeenCalledWith({
          message: JSON.stringify(createdCourse),
          partition: null,
          topic: "new-course",
        })

        expect(createdCourse).not.toEqual(null)
        expect(createdCourse!.id).toEqual(res.addCourse.id)
        expect(createdCourse).toMatchSnapshot({
          created_at: expect.any(Date),
          updated_at: expect.any(Date),
          id: expect.any(String),
          photo_id: (expected as any).photo ? expect.any(String) : null,
        })
      })

      it("errors with non-admin", async () => {
        ctx!.client.setHeader("Authorization", "Bearer normal")
        try {
          await ctx!.client.request(createCourseMutation, {
            course: cases[0],
          })
          fail()
        } catch {}
      })
    })

    describe("updateCourse", () => {
      beforeEach(async () => {
        await seed(ctx.prisma)
        ctx!.client.setHeader("Authorization", "Bearer admin")
      })

      it("errors on no slug given", async () => {
        try {
          await ctx!.client.request(updateCourseMutation, {
            course: omit(getUpdateCourse(), "slug"),
          })
          fail()
        } catch {}
      })

      it("updates course", async () => {
        const res = await ctx!.client.request(updateCourseMutation, {
          course: {
            ...getUpdateCourse(),
            new_photo: undefined,
          },
        })

        expect(res).toMatchSnapshot({
          updateCourse: {
            ...expectedUpdatedCourse,
            photo: {
              ...expectedUpdatedCourse.photo,
              original: expect.stringContaining("original.gif"),
            },
          },
        })

        const oldCourse = await ctx.prisma.course.findFirst({
          where: { slug: "course1" },
        })
        expect(oldCourse).toBeNull()
        const updatedCourse = await ctx.prisma.course.findFirst({
          where: { slug: "updated_course1" },
        })
        expect(updatedCourse).toMatchSnapshot({
          created_at: expect.any(Date),
          updated_at: expect.any(Date),
          photo_id: expect.any(String),
        })
      })

      it("updates photo", async () => {
        const res = await ctx!.client.request(updateCourseMutation, {
          course: getUpdateCourse(),
        })

        expect(res).toMatchSnapshot({
          updateCourse: expectedUpdatedCourse,
        })

        const oldCourse = await ctx.prisma.course.findFirst({
          where: { slug: "course1" },
        })
        expect(oldCourse).toBeNull()
        const oldImage = await ctx.prisma.image.findFirst({
          where: { id: "00000000000000000000000000001101" },
        })
        expect(oldImage).toBeNull()

        const updatedCourse = await ctx.prisma.course.findFirst({
          where: { slug: "updated_course1" },
        })
        expect(updatedCourse).toMatchSnapshot({
          created_at: expect.any(Date),
          updated_at: expect.any(Date),
          photo_id: expect.any(String),
        })
      })

      it("deletes photo", async () => {
        const res = await ctx!.client.request(updateCourseMutation, {
          course: {
            ...getUpdateCourse(),
            new_photo: undefined,
            new_slug: undefined,
            delete_photo: true,
          },
        })

        expect(res).toMatchSnapshot({
          updateCourse: {
            ...expectedUpdatedCourse,
            photo: null,
          },
        })
        const updatedCourse = await ctx.prisma.course.findFirst({
          where: { slug: "course1" },
        })
        expect(updatedCourse).toMatchSnapshot({
          created_at: expect.any(Date),
          updated_at: expect.any(Date),
        })
      })

      it("errors with non-admin", async () => {
        ctx!.client.setHeader("Authorization", "Bearer normal")
        try {
          await ctx!.client.request(updateCourseMutation, {
            course: getUpdateCourse(),
          })
          fail()
        } catch {}
      })
    })

    describe("deleteCourse", () => {
      beforeEach(async () => {
        await seed(ctx.prisma)
        ctx!.client.setHeader("Authorization", "Bearer admin")
      })

      it("deletes course on id", async () => {
        const res = await ctx!.client.request(deleteCourseMutation, {
          id: "00000000000000000000000000000002",
        })

        expect(res).toMatchSnapshot()
        const deletedCourse = await ctx.prisma.course.findFirst({
          where: { id: "00000000000000000000000000000002" },
        })
        expect(deletedCourse).toBeNull()
        const deletedImage = await ctx.prisma.image.findFirst({
          where: { id: "00000000000000000000000000001101" },
        })
        expect(deletedImage).toBeNull()
      })

      it("deletes course on slug", async () => {
        const res = await ctx!.client.request(deleteCourseMutation, {
          slug: "course1",
        })

        expect(res).toMatchSnapshot()
        const deletedCourse = await ctx.prisma.course.findFirst({
          where: { slug: "course1" },
        })
        expect(deletedCourse).toBeNull()
      })

      it("errors with non-admin", async () => {
        ctx!.client.setHeader("Authorization", "Bearer normal")
        try {
          await ctx!.client.request(deleteCourseMutation, {
            slug: "course1",
          })
          fail()
        } catch {}
      })
    })
  })
})
