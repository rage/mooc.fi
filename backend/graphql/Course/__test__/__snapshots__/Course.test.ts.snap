// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Course model completions shouldn't return anything with non-existent user 1`] = `
Object {
  "course": Object {
    "completions": Array [],
    "id": "00000000-0000-0000-0000-000000000002",
    "name": "course1",
    "slug": "course1",
  },
}
`;

exports[`Course model completions works with user_id 1`] = `
Object {
  "course": Object {
    "completions": Array [
      Object {
        "completion_language": null,
        "email": "what@ever.com",
        "id": "12400000-0000-0000-0000-000000000001",
        "user": {
          "id": "20000000-0000-0000-0000-000000000103",
          "username": "admin",
        },
        "user_upstream_id": null,
      },
    ],
    "id": "00000000-0000-0000-0000-000000000002",
    "name": "course1",
    "slug": "course1",
  },
}
`;

exports[`Course model completions works with user_upstream_id 1`] = `
Object {
  "course": Object {
    "completions": Array [
      Object {
        "completion_language": null,
        "email": "e@mail.com",
        "id": "30000000-0000-0000-0000-000000000102",
        "user": {
          "id": "20000000-0000-0000-0000-000000000102",
          "username": "existing_user",
        },
        "user_upstream_id": 1,
      },
    ],
    "id": "00000000-0000-0000-0000-000000000002",
    "name": "course1",
    "slug": "course1",
  },
}
`;
